RecoPixelVertexing/PixelTriplets/plugins/gpuPixelDoublets.h

```diff
@@ -11,151 +11,162 @@
-    auto i = blockIdx.x*blockDim.x + threadIdx.x;
-    if (i>=offsets[9]) {
-      // get rid of last layer
-      return;
-    }
+    constexpr float z0cut = 12.f;
+    auto z0cutoff = [&](int j) {
+      auto zo =	hh.zg_d[j];
+      auto ro = hh.rg_d[j]; 
+      auto dr = ro-mer;
+      return dr > maxr[pairLayerId] || 
+             dr<0 || std::abs((mez*ro - mer*zo)) > z0cut*dr;
+    };
 
-    assert(0==offsets[0]);
-    int top = (i>offsets[5]) ? 5: 0;
-    while (i>=offsets[++top]){};
-    assert(top<10);
-    auto bottom = top-1;
-    if (bottom==3 || bottom==6) {
-      // do not have UP... (9 we got rid already)
-      return;
-    }
-    assert(i>=offsets[bottom]);
-    assert(i<offsets[top]);
+    auto iphicut = phicuts[pairLayerId];
 
-    auto mep = iphi[i];
-    auto kl = hist[top].bin(mep-iphicut);
-    auto kh = hist[top].bin(mep+iphicut);
+    auto kl = hist[outer].bin(int16_t(mep-iphicut));
+    auto kh = hist[outer].bin(int16_t(mep+iphicut));
     auto incr = [](auto & k) { return k = (k+1)%Hist::nbins();};
     int tot  = 0;
     int nmin = 0;
     auto khh = kh;
     incr(khh);
+    
+    int tooMany=0;
     for (auto kk=kl; kk!=khh; incr(kk)) {
-      if (kk!=kl && kk!=kh) nmin+=hist[top].size(kk);
-      for (auto p=hist[top].begin(kk); p<hist[top].end(kk); ++p) {
-        if (std::min(std::abs(int16_t(iphi[*p]-mep)), std::abs(int16_t(mep-iphi[*p]))) > iphicut)
+      if (kk!=kl && kk!=kh) nmin+=hist[outer].size(kk);
+      for (auto p=hist[outer].begin(kk); p<hist[outer].end(kk); ++p) {
+        auto oi=*p;
+        assert(oi>=offsets[outer]);
+        assert(oi<offsets[outer+1]);
+
+        if (std::min(std::abs(int16_t(iphi[oi]-mep)), std::abs(int16_t(mep-iphi[oi]))) > iphicut)
           continue;
+        if (z0cutoff(oi)) continue;
+        auto ind = atomicInc(nCells,MaxNumOfDoublets);
+        // int layerPairId, int doubletId, int innerHitId,int outerHitId)
+        cells[ind].init(hh,pairLayerId,ind,i,oi);
+        isOuterHitOfCell[oi].push_back(ind);
+        if (isOuterHitOfCell[oi].full()) ++tooMany;
         ++tot;
       }
     }
-    if (0==hist[top].nspills) assert(tot>=nmin);
+    if (tooMany>0) printf("OuterHitOfCell full for %d in layer %d/%d, %d:%d   %d,%d\n", i, inner,outer, kl,kh,nmin,tot);
+
+    if (hist[outer].nspills>0)
+      printf("spill bin to be checked in %d %d\n",outer,hist[outer].nspills);
+
+    // if (0==hist[outer].nspills) assert(tot>=nmin);
     // look in spill bin as well....
-  }
 
+
+  }  // loop in block...
+  }
   __global__
-  void getDoubletsFromHisto(siPixelRecHitsHeterogeneousProduct::HitsOnGPU const * hhp, float phiCut) {
+  void getDoubletsFromHisto(GPUCACell * cells, uint32_t * nCells, siPixelRecHitsHeterogeneousProduct::HitsOnGPU const * hhp,                
+                            GPU::VecArray< unsigned int, 2048> *isOuterHitOfCell) {
+
+    uint8_t const layerPairs[2*13] = {0,1 ,1,2 ,2,3 
+                                     // ,0,4 ,1,4 ,2,4 ,4,5 ,5,6  
```

Comment:
a test is two switch forward and backward (again).
if results do not reproduce than there is a problem